src/Storages/NATS/NATSLibUVAdapter.cpp

```diff
@@ -0,0 +1,406 @@
+/**
+*  Patched adapter from NATS io library to plug a `NATS` connection to a `libuv` event loop.
+ */
+
+#include <Storages/NATS/NATSLibUVAdapter.h>
+
+static uv_once_t    uvOnce = UV_ONCE_INIT;
+static uv_key_t     uvLoopThreadKey;
+
+static void initOnce()
+{
+    if (uv_key_create(&uvLoopThreadKey) != 0)
+        abort();
+}
+
+void natsLibuvInit()
+{
+    uv_once(&uvOnce, initOnce);
+}
+
+void natsLibuvSetThreadLocalLoop(uv_loop_t *loop)
+{
+    uv_key_set(&uvLoopThreadKey, static_cast<void*>(loop));
+}
+
+/// Patched uvScheduleToEventLoop function from NATS io libuv adapter (contrib/nats-io/src/adapters/libuv.h)
+static natsStatus uvScheduleToEventLoop(NATSLibuvEvents *nle, int eventType, bool add)
+{
+    NATSLibuvEvent  *newEvent = nullptr;
+    int             res;
+
+    uv_mutex_lock(nle->lock);
+    if (!nle->scheduler)
+    {
+        uv_mutex_unlock(nle->lock);
+        return NATS_ILLEGAL_STATE;
+    }
+    uv_mutex_unlock(nle->lock);
+
+    newEvent = static_cast<NATSLibuvEvent*>(malloc(sizeof(NATSLibuvEvent)));
+    if (newEvent == nullptr)
+        return NATS_NO_MEMORY;
+
+    newEvent->type  = eventType;
+    newEvent->add   = add;
+    newEvent->next  = nullptr;
+
+    uv_mutex_lock(nle->lock);
+
+    if (nle->head == nullptr)
+        nle->head = newEvent;
+
+    if (nle->tail != nullptr)
+        nle->tail->next = newEvent;
+
+    nle->tail = newEvent;
+
+    res = uv_async_send(nle->scheduler);
+
+    uv_mutex_unlock(nle->lock);
+
+    return (res == 0 ? NATS_OK : NATS_ERR);
+}
+
+static void natsLibuvPoll(uv_poll_t* handle, int status, int events)
+{
+    NATSLibuvEvents *nle = static_cast<NATSLibuvEvents*>(handle->data);
+
+    if (status != 0)
+    {
+        // There was an error, try to process as a read event.
+        // If we had an issue with the socket, this will cause
+        // an auto-reconnect.
+        natsConnection_ProcessReadEvent(nle->nc);
+        return;
+    }
+
+    if (events & UV_READABLE)
+        natsConnection_ProcessReadEvent(nle->nc);
+
+    if (events & UV_WRITABLE)
+        natsConnection_ProcessWriteEvent(nle->nc);
+}
+
+static natsStatus uvPollUpdate(NATSLibuvEvents *nle, int eventType, bool add)
+{
+    int res;
+
+    if (eventType == NATS_LIBUV_READ)
+    {
+        if (add)
+            nle->events |= UV_READABLE;
+        else
+            nle->events &= ~UV_READABLE;
+    }
+    else
+    {
+        if (add)
+            nle->events |= UV_WRITABLE;
+        else
+            nle->events &= ~UV_WRITABLE;
+    }
+
+    if (nle->events)
+        res = uv_poll_start(nle->handle, nle->events, natsLibuvPoll);
+    else
+        res = uv_poll_stop(nle->handle);
+
+    if (res != 0)
+        return NATS_ERR;
+
+    return NATS_OK;
+}
+
+static void uvHandleClosedCb(uv_handle_t *handle)
+{
+    free(handle);
```

Comment:
Also these C like memory handling (malloc, calloc, free, etc) should be replaced by their C++ counterpart.